app/backend/src/couchers/jobs/worker.py

```diff
@@ -0,0 +1,183 @@
+"""
+Background job workers
+"""
+
+import logging
+import traceback
+from concurrent import futures
+from datetime import timedelta
+from multiprocessing import Process
+from sched import scheduler
+from time import monotonic, sleep
+
+from google.protobuf import empty_pb2
+from sqlalchemy.sql import func
+
+from couchers.db import get_engine, session_scope
+from couchers.jobs.definitions import JOBS, SCHEDULE
+from couchers.jobs.enqueue import queue_job
+from couchers.models import BackgroundJob, BackgroundJobState, BackgroundJobType, RepeatedJob
+from couchers.utils import now
+
+logger = logging.getLogger(__name__)
+
+
+def process_job():
+    logger.info(f"Looking for a job")
+    with session_scope(isolation_level="REPEATABLE READ") as session:
+        # a combination of REPEATABLE READ and SELECT ... FOR UPDATE SKIP LOCKED makes sure that only one transaction
+        # will modify the job at a time. SKIP UPDATE means that if the job is locked, then we ignore that row, it's
+        # easier to use SKIP LOCKED vs NOWAIT in the ORM, with NOWAIT you get an ugly exception from deep inside
+        # psycopg2 that's quite annoying to catch and deal with
+        job = (
+            session.query(BackgroundJob).filter(BackgroundJob.ready_for_retry).with_for_update(skip_locked=True).first()
+        )
+
+        if not job:
+            logger.info(f"No pending jobs")
+            return
+
+        # we've got a lock for a job now, it's "pending" until we commit or the lock is gone
+        logger.info(f"Job #{job.id} grabbed")
+
+        job.try_count += 1
+
+        message_type, func = JOBS[job.job_type]
+
+        try:
+            ret = func(message_type.FromString(job.payload))
+            job.state = BackgroundJobState.completed
+            logger.info(f"Job #{job.id} complete on try number {job.try_count}")
+        except Exception as e:
+            logger.exception(e)
+            if job.try_count >= job.max_tries:
+                # if we already tried max_tries times, it's permanently failed
+                job.state = BackgroundJobState.failed
+                logger.info(f"Job #{job.id} failed on try number {job.try_count}")
+            else:
+                job.state = BackgroundJobState.error
+                # exponential backoff
+                job.next_attempt_after += timedelta(seconds=15 * (2 ** job.try_count))
+                logger.info(f"Job #{job.id} error on try number {job.try_count}, next try at {job.next_attempt_after}")
+            # add some info for debugging
+            job.failure_info = traceback.format_exc()
+
+        # exiting ctx manager commits and releases the row lock
+
+
+def service_jobs():
+    """
+    Don't run many of these simultaneously: it'll work but it'll conflict and cause a lot of missed job processing.
+    """
```

Comment:
Why? The row locking should be solid.

Reply:
Yeah, row locking itself is solid in that jobs themselves will be run only once. But the scheduler checks how many jobs there are and launches a worker for each. If there's multiple schedulers they might launch twice the workers causing half the workers to not acquire a job and exit. Just unnecessarily burdening the DB.

Of course this depends only really on the amount of time it takes to check how many pending jobs there are vs how that worker starting and attempting to acquire a job. If that time is very small (which it should really be), and there are multiple schedulers, they shouldn't really interfere with each other.

Reply:
If the workers threads are slow to start working on the ThreadPoolExecutor submitted jobs, then the main thread will keep submitting new jobs to the ThreadPoolExecutor, possibly making a bad situation worse.

This is unnecessary complex IMHO. Just start all the workers immediately and let them poll the database themselves.

Reply:
Hmmm, that's a good point.

I was hoping to eventually move to a pg `LISTEN ... NOTIFY` system to spawn workers based on actual events, and run the scheduler only to clean up any jobs that were left over in case. That's what inspired this style.

I agree it's more complex now, but I think now it's pretty solid. It doesn't really matter now though. I can simplify it though if you think it's better. (I would prefer to keep `process_job` mostly intact though, and just have the `service_jobs` function call it directly, waiting a second if it got no job)

Reply:
> I would prefer to keep process_job mostly intact though, and just have the service_jobs function call it directly, waiting a second if it got no job

This sounds good. `process_job` is good to keep as a function, releasing the lock automatically when it returns or throw exception.

Reply:
Simplified, what do you think?

Reply:
Much better! We might need to adjust the log levels of the spamming log messages but that's details. Resolving conversation!